File: memento/go/memento.go
package memento

import "bytes"

type Snapshot struct {
	Bakpoint int
}

type InputText struct {
	text bytes.Buffer
}

func (t *InputText) GetText() string {
	return t.text.String()
}

func (t *InputText) Append(input string) {
	t.text.WriteString(input)
}

func (t *InputText) CreateSnapshot() *Snapshot {
	return &Snapshot{t.text.Len()}
}

func (t *InputText) RestoreSnapshot(snapshot *Snapshot) {
	t.text.Truncate(snapshot.Bakpoint)
}

type SnapshotHolder struct {
	snapshots []*Snapshot
}

func (h *SnapshotHolder) PopSnapshot() *Snapshot {
	if len(h.snapshots) == 0 {
		return nil
	}
	v := h.snapshots[len(h.snapshots)-1]
	h.snapshots = h.snapshots[:len(h.snapshots)-1]
	return v
}

func (h *SnapshotHolder) PushSnapshot(snapshot *Snapshot) {
	if len(h.snapshots) >= 3 {
		h.snapshots = h.snapshots[1:]
	}
	h.snapshots = append(h.snapshots, snapshot)
}

File: memento/java/Memento.java
import java.util.Deque;
import java.util.concurrent.ConcurrentLinkedDeque;

record Snapshot(int bakpoint) {}

class InputText {
    private final StringBuilder text = new StringBuilder();

    public String getText() {
        return text.toString();
    }

    public void append(String input) {
        text.append(input);
    }

    public Snapshot createSnapshot() {
        return new Snapshot(text.length());
    }

    public void restoreSnapshot(Snapshot snapshot) {
        text.delete(snapshot.bakpoint(), text.length());
    }
}

class SnapshotHolder {
    // Stack 使用栈结构
    private final Deque<Snapshot> snapshots = new ConcurrentLinkedDeque<>();

    public Snapshot popSnapshot() {
        if (snapshots.isEmpty()) {
            return null;
        }
        return snapshots.pop();
    }

    public void pushSnapshot(Snapshot snapshot) {
        if (snapshots.size() >= 3) {
            snapshots.removeLast();
        }
        snapshots.push(snapshot);
    }
}